🗓️ Fixed WSOD on Metric Date Selector

the-bay-kay opened the pull request:
Fix for the WSOD that was reported in issue [1027](https://github.com/e-mission/e-mission-docs/issues/1027) .  Now, when a single day is picked, the range queried is "the full day" of the start date.  

This was an issue on both iOS and Android: I've tested this fix on iOS, but am currently unable to test on Android.  For anyone able to, please let me know if it works on both OSs! 

(PR is to merge into @JGreenlee 's UI wrapup, since it's the most contemporary branch!).

https://github.com/e-mission/e-mission-phone/assets/98350084/4b7ca1c9-59c3-45ad-8bd1-7824e74ddae4

the-bay-kay committed:
Fixed WSOD on single-date pick

the-bay-kay committed:
Clean up a line of test code

JGreenlee commented on www/js/metrics/MetricsDateSelect.tsx:
```diff
@@ -35,10 +35,18 @@ const MetricsDateSelect = ({ dateRange, setDateRange }: Props) => {
+      const dtStartDate = DateTime.fromJSDate(startDate).startOf('day')
+      let dtEndDate;
+      if(!endDate) { // For when only one day is selected
+        dtEndDate = dtStartDate.endOf('day').minus({minutes: 1});
```
Is there a good reason to use the endOf the day here when usually we just use the startOf the day?
If we used startOf the day for both ends of the selected range, is the behavior any different?

I am not sure that the server cares what part of the day the request timestamps refer to since it chunks into full days anyway -- but let me know if you noticed otherwise.
If there is no difference, this code can be simplified a lot

the-bay-kay replied:
Gosh, thank you for asking -- This is going to be a wall of text, but TL;DR, I think this is a lot more subtle of an issue than I originally thought!

First, I think we probably don't need the `minus()` -- I remember it making a difference and decided to keep it, but can't seem to replicate that again... I'm thinking I'll test a little more, and axe it if I don't see a difference.

If we set the `endDate` to the `startDate.startOf()`, the server query returns blank. I had the same idea, since (irrc) BEMUserCache queries work fine with a single timestamp.  From what understand, the `BEMServerComm` to query to "result/metrics/${}"  in commHelper needs the times to be different.  I haven't looked too far into it, but will do so after writing this up.

Interestingly, this brings me to another issue I just found as I was writing this -- there actually _is_ a way to select a single date range, without encountering the `null` value.  This is done, somewhat unintuitively, by double tapping given date.  But, because both values are the start-date, this  shows up empty as mentioned above.

I also noticed another subtle issue when working with `endOf()` or `endOf().minus()`. The dateRange on the top right shows the correct "range", however the `MyFootprint` card suggests that it's a two day range (in the videos, 12/4-12/5).  Upon checking that two-day range, the metrics _are_ the same. 

So,  this current commit "fixed" the WSOD, and made it possible to select one day -- but the behavior isn't exactly what we want, since it technically shows the range of two days.  I think there's a couple things I want to try:
- Make the query _actually_ one day; from basic testing, it doesn't behave like the other queries (where one timestamp is sufficient), so I want to look into how we could pull a single day
- If we _can't_ pull the single day, I want to brainstorm ways to prompt the user to re-select.  It seems disingenuous to leave this as is, since the info displayed isn't exactly what the user asked for.  I remember reading that the react-native calendar doesn't have a way to prevent a single date from being picked, but I bet we could prompt the user another way (Make the blank background say "Please select two or more days!", or something like that...)

I'll spend some time staring at the server code, and report back later!

## Old "Null" Error ↓
https://github.com/JGreenlee/e-mission-phone/assets/98350084/8d8473fc-6438-443d-953c-d8d24244fa97
## Start Date Returns Blank ↓
https://github.com/JGreenlee/e-mission-phone/assets/98350084/ff1a73a0-6c98-4ff2-82fb-4f492d4dadd4
## "Proper" Single Day Select Returns Blank ↓
https://github.com/JGreenlee/e-mission-phone/assets/98350084/629413ba-4b93-40bc-a1ad-c08bdafb1d0e

## Single Day == Two Day ↓
https://github.com/JGreenlee/e-mission-phone/assets/98350084/6d75d9c5-89a3-4817-8671-f3e0551eabdc

the-bay-kay replied:
Just a quick update on this problem:

Below is, I believe, the call trace for fetching / collecting this data. Trying to organize this trace to I can come back to it: each row has the function, and where it's called (as such, the next row down has the location of the called function's definition).  First two calls are from the client, the rest are from the server.

| Function | Func Call Location ↓ | Notes |
|--------------|------------------|-----|
| GetMetrics() | `MetricsTab.jsx `  | Metrics are sorted as a state |
| "BEMServer fetch"  |  `commHelper.ts` | Posts to server | 
| summarize_by_timestamp() | `cfc_webapp.py`| '/result/metrics/<time_type>'|
| _group_by_timestamp() |  `metrics.py` | Called as HOF in _call_group_fn() |
| get_data_df() |  `time_grouping.py` | DF also used elsewhere| 
| get_time_series() | `analysis_timeseries_queries.py` | called from get_timeseries_for_user |
| BuiltinTimeSeries() | `abstract_timeseries.py` | Abstract calls base Constructor Call |
| _get_query() | `builtin_timeseries.py` | I believe _this_ is fetches from the DB !|

With all of this laid out, I've been struggling to find exaclty _why_ we aren't able to fetch with a single timestamp.  If I understand the stack correctly, the `builtin_timeseries` objects should be able to hold hold single days.  I've got a lot more digging to do!

the-bay-kay committed:
Added to WSOD fix on metrics tab

When selecting date, the following behavior occurs:
- If 2+ days, behave as normal (Range of Day1 -> Day2)
- If  "open Ended" (i.e., only start is picked), fetch
  a range of days between that day and present (Day1 -> Today)
- If 1 day, fetch range for 48 hours from that day

TODO: The 1 day range going to 48 hours is meant as a patch, so that
this WSOD fix can go to staging.  To make a single day only fetch
24 hrs, some changes to the server may be needed.

the-bay-kay commented:
Made some tweaks, ran prettier, this should be good to go! Here's a few videos of the fix running, and  the edge cases.

**Future Work:** As written, the fetch for a single day _technically_ fetches 48 hours from the start of the selected date. After some discussion, we agreed this is fine for now -- the focus of this PR was to fix the WSOD.  We also tested the day-of  fetch, and confirmed it will not crash when querying the current date (i.e., data missing from "tomorrow" doesn't cause problems in the 48-hour fetch).

### Normal Range (2+ Days)

https://github.com/JGreenlee/e-mission-phone/assets/98350084/b97e59f1-0d76-4e0f-8bed-1558f11ff7dc

### Open Ended 

https://github.com/JGreenlee/e-mission-phone/assets/98350084/d1c59c6f-0e37-4e3b-8c35-14c61190f64b

### Single Day

https://github.com/JGreenlee/e-mission-phone/assets/98350084/42acc7b4-932e-49c6-99e9-e4b980ab2bfc

JGreenlee reviewed (approved):
Great work investigating and implementing a solution here!
